server/lib/wsapi/matchmaking.js

```diff
@@ -138,50 +155,119 @@ export class MatchmakingApi {
 
   gameLoaderDelegate = {
     onGameSetup: async (matchInfo, clients, players, setup = {}) => {
-      // TODO(2Pac): Select map intelligently based on user's preference
-      const mapPool = await getCurrentMapPool(matchInfo.type)
-      if (!mapPool) {
+      const currentMapPool = await getCurrentMapPool(matchInfo.type)
+      if (!currentMapPool) {
         throw new Error('invalid map pool')
       }
 
-      const mapInfo = (await getMapInfo(mapPool.maps))[0]
-      if (!mapInfo) {
-        throw new Error('invalid map')
+      const mapPool = new Set(currentMapPool.maps)
+      const preferredMapsHashes = matchInfo.players
+        .reduce((acc, p) => acc.concat(p.preferredMaps), new Set())
+        .filter(m => mapPool.includes(m))
+
+      const randomMapsHashes = []
+      Range(preferredMapsHashes.size, 4).forEach(() => {
+        const availableMaps = mapPool.subtract(preferredMapsHashes.concat(randomMapsHashes))
+        const randomMap = availableMaps.toList().get(getRandomInt(availableMaps.size))
+        randomMapsHashes.push(randomMap)
+      })
+
+      const [preferredMaps, randomMaps] = await Promise.all([
+        await getMapInfo(preferredMapsHashes.toJS()),
```

Comment:
since you're awaiting the outer Promise, you don't want to await these inner ones (or you're ending up blocking on them here anyway)